server/src/main/java/com/objectcomputing/checkins/services/checkins/CheckInServicesImpl.java

```diff
@@ -1,86 +1,167 @@
+                // Limit read to Subject of check-in, PDL of subject and Admin
+                throw new CheckInBadArgException(String.format("Member %s is unauthorized to do this operation", currentUser.getId()));
+            }
+        }
+
+        return result;
     }
 
     @Override
-    public CheckIn update(CheckIn checkIn) {
-        CheckIn checkInRet = null;
-        if(checkIn!=null){
+    public CheckIn update(@NotNull CheckIn checkIn) {
+
         final UUID id = checkIn.getId();
         final UUID memberId = checkIn.getTeamMemberId();
-        LocalDate chkInDate = checkIn.getCheckInDate();
-        if(id==null||!checkinRepo.findById(id).isPresent()){
+        final UUID pdlId = checkIn.getPdlId();
+        LocalDateTime chkInDate = checkIn.getCheckInDate();
+
+        String workEmail = securityService!=null ? securityService.getAuthentication().get().getAttributes().get("email").toString() : null;
+        MemberProfile currentUser = workEmail!=null? currentUserServices.findOrSaveUser(null, workEmail) : null;
+        Boolean isAdmin = securityService!=null ? securityService.hasRole(RoleType.Constants.ADMIN_ROLE) : false;
+
+        if(id==null||!checkinRepo.findById(id).isPresent()) {
             throw new CheckInBadArgException(String.format("Unable to find checkin record with id %s", checkIn.getId()));
-        }else if(!memberRepo.findById(memberId).isPresent()){
-            throw new CheckInBadArgException(String.format("Member %s doesn't exist", memberId));
         } else if(memberId==null) {
             throw new CheckInBadArgException(String.format("Invalid checkin %s", checkIn));
-        } else if(chkInDate.isBefore(LocalDate.EPOCH) || chkInDate.isAfter(LocalDate.MAX)) {
+        } else if(!memberRepo.findById(memberId).isPresent()) {
+            throw new CheckInBadArgException(String.format("Member %s doesn't exist", memberId));
+        } else if(!pdlId.equals(memberRepo.findById(memberId).get().getPdlId())) {
+            throw new CheckInBadArgException(String.format("PDL %s is not associated with member %s", pdlId, memberId));
+        } else if(chkInDate.isBefore(Util.MIN) || chkInDate.isAfter(Util.MAX)) {
             throw new CheckInBadArgException(String.format("Invalid date for checkin %s",memberId));
+        } else if(!isAdmin) {
+            if(!currentUser.getId().equals(checkIn.getTeamMemberId()) && !currentUser.getId().equals(checkIn.getPdlId())) {
+                // Limit update to subject of check-in, PDL of subject and Admin
+                throw new CheckInBadArgException(String.format("Member %s is unauthorized to do this operation", currentUser.getId()));
+            } else if(checkinRepo.findById(id).get().isCompleted()) {
+                // Update is only allowed if the check in is not completed unless made by admin
+                throw new CheckInBadArgException(String.format("Checkin with id %s is complete and cannot be updated", checkIn.getId()));
+            }
```

Comment:
I feel like this could be made more concise by defining a method that would take care of the `if...throw` bit maybe?

It could take a lambda, a message, and an optional set of replacement values...or even a boolean instead of a lambda.  There's no need for the elses since an exception being thrown...so it could just be a sequence of validation calls. Just a thought.

Reply:
I've added a private method called validate in checkinServiceImpl in the following commit - https://github.com/oci-labs/check-ins/pull/283/commits/7437ac3545ee2026d01ec738582b3e7ee7c20a08